cast/src/lib.rs

```diff
@@ -1656,6 +1662,75 @@ impl SimpleCast {
     pub fn disassemble(bytecode: &str) -> Result<String> {
         format_operations(disassemble_str(bytecode)?)
     }
+
+    /// Gets the selector for a given function signature
+    /// Optimizes if the `optimize` parameter is set to a number of leading zeroes
+    ///
+    /// # Example
+    ///
+    /// ```
+    /// use cast::SimpleCast as Cast;
+    ///
+    /// fn main() -> eyre::Result<()> {
+    ///     assert_eq!(Cast::get_selector(String::from("foo(address,uint256)"), None)?.0, String::from("0xbd0d639f"));
+    ///     assert_eq!(Cast::get_selector(String::from("foo(address,uint256)"), Some(2))?.0, String::from("0x0000bf3e"));
+    ///     assert_eq!(Cast::get_selector(String::from("foo(address,uint256)"), Some(2))?.1, Some(String::from("foo71661(address,uint256)")));
+    ///
+    ///     Ok(())
+    /// }
+    /// ```    
+    pub fn get_selector(
+        signature: String,
+        optimize: Option<usize>,
+    ) -> Result<(String, Option<String>, Option<Duration>)> {
+        if optimize.is_none() || optimize.unwrap() > 4 {
+            let selector = HumanReadableParser::parse_function(&signature)?.short_signature();
+            return Ok((format!("0x{}", hex::encode(selector)), None, None))
+        }
+        let mut name: &str = "";
+        let mut params: &str = "";
+        if let Some(index) = signature.find('(') {
+            name = &signature[..index];
+            params = &signature[index..];
+        }
+
+        let num_threads = num_cpus::get();
+        let start_time = Instant::now();
+        let found = AtomicBool::new(false);
+
+        let results: Vec<Option<(u32, String, String)>> = (0..num_threads)
+            .into_par_iter()
+            .map(|i| {
+                let nonce_start = i as u32;
+                let nonce_step = num_threads as u32;
+
+                let mut nonce = nonce_start;
+                while nonce < std::u32::MAX && !found.load(Ordering::Relaxed) {
+                    let input = format!("{}{}{}", name, nonce, params);
+                    let hash = keccak256(input.as_bytes());
+                    let selector = &hash[..4];
+
+                    if selector.iter().take(optimize.unwrap()).all(|&byte| byte == 0) {
```

Comment:
This can be optimized further by using [`u32::from_be_bytes()`](https://doc.rust-lang.org/stable/std/primitive.u32.html#method.from_be_bytes) and [`u32::leading_zeroes()`](https://doc.rust-lang.org/stable/std/primitive.u32.html#method.leading_zeros)

Something like this

```suggestion
                    let selector = u32::from_be_bytes(hash[..4].try_into().unwrap());

                    if selector.leading_zeroes() == optimize {
```

Reply:
this requires casting the `selector` from u32 to [u8]

Reply:
Actually no that wouldn't work since we want `0` bytes not bits..

Here's a few faster alternatives: <https://godbolt.org/z/ssGM87dT3>

Just a nit at this point